JVM compile errors are not propagated in any case

Javac errors in generated code (protobuf in this example) are not propagated to be visible to the user. I used example-codegen with https://github.com/pantsbuild/example-codegen/pull/10 to test compilation of Java Protobuf code. (You will need to disable Python and Scala backends in the PR branch.)

Results in:

```
18:20:42.77 [INFO] Starting new nailgun server with cmd: "/bin/bash", args ["__jdk/jdk.sh", "__java_home/bin/java", "-cp", "__jdk/com.martiansoftware_nailgun-server_0.9.1.jar:__java_home/lib/tools.jar", "com.martiansoftware.nailgun.NGServer", ":0"], in cwd /private/var/folders/md/0q71p41n0rbgwhnc8npjjzg00000gn/T/process-executionzT7L2w
18:20:43.95 [ERROR] Completed: Check javac compilation - javac failed (exit code 1).

✕ javac failed.
```

No relevant output as to why the compile failed was emitted.

--- a/src/rust/engine/process_execution/src/cache.rs
+++ b/src/rust/engine/process_execution/src/cache.rs
@@ -88,10 +88,12 @@ impl crate::CommandRunner for CommandRunner {
             }
             // When we successfully use the cache, we change the description and increase the level
             // (but not so much that it will be logged by default).
-            workunit.update_metadata(|initial| WorkunitMetadata {
-              desc: initial.desc.as_ref().map(|desc| format!("Hit: {}", desc)),
-              level: Level::Debug,
-              ..initial
+            workunit.update_metadata(|initial| {
+              initial.map(|initial| WorkunitMetadata {
+                desc: initial.desc.as_ref().map(|desc| format!("Hit: {}", desc)),
+                level: Level::Debug,
+                ..initial
+              })
             });
             Ok(result)
           }

--- a/src/rust/engine/process_execution/src/remote.rs
+++ b/src/rust/engine/process_execution/src/remote.rs
@@ -808,13 +808,15 @@ impl crate::CommandRunner for CommandRunner {
         let result_fut = self.run_execute_request(execute_request, request, &context2, workunit);
         let result = tokio::time::timeout(deadline_duration, result_fut).await;
         if result.is_err() {
-          workunit.update_metadata(|inititial| WorkunitMetadata {
-            level: Level::Error,
-            desc: Some(format!(
-              "remote execution timed out after {:?}",
-              deadline_duration
-            )),
-            ..inititial
+          workunit.update_metadata(|initial| {
+            Some(WorkunitMetadata {
+              level: Level::Error,
+              desc: Some(format!(
+                "remote execution timed out after {:?}",
+                deadline_duration
+              )),
+              ..initial.unwrap_or_default()
+            })
           })
         }
 

--- a/src/rust/engine/process_execution/src/remote_cache.rs
+++ b/src/rust/engine/process_execution/src/remote_cache.rs
@@ -298,14 +298,17 @@ impl CommandRunner {
               }
               // When we successfully use the cache, we change the description and increase the level
               // (but not so much that it will be logged by default).
-              workunit.update_metadata(|initial| WorkunitMetadata {
-                desc: initial
-                  .desc
-                  .as_ref()
-                  .map(|desc| format!("Hit: {}", desc)),
-                level: Level::Debug,
-                ..initial
+              workunit.update_metadata(|initial| {
+                initial.map(|initial|
+                WorkunitMetadata {
+                  desc: initial
+                    .desc
+                    .as_ref()
+                    .map(|desc| format!("Hit: {}", desc)),
+                  level: Level::Debug,
+                  ..initial
 
+                })
               });
               Ok((cached_response, true))
             } else {

--- a/src/rust/engine/src/externs/engine_aware.rs
+++ b/src/rust/engine/src/externs/engine_aware.rs
@@ -18,32 +18,30 @@ use workunit_store::{ArtifactOutput, Level, RunningWorkunit, UserMetadataItem};
 // be like the user did not set extra metadata.
 
 #[derive(Default, Clone, Debug)]
-pub(crate) struct EngineAwareReturnType {
-  level: Option<Level>,
-  message: Option<String>,
-  metadata: Vec<(String, UserMetadataItem)>,
-  artifacts: Vec<(String, ArtifactOutput)>,
-}
+pub(crate) struct EngineAwareReturnType;
 
 impl EngineAwareReturnType {
-  pub(crate) fn from_task_result(task_result: &PyAny) -> Self {
-    Self {
-      level: Self::level(task_result),
-      message: Self::message(task_result),
-      artifacts: Self::artifacts(task_result).unwrap_or_else(Vec::new),
-      metadata: metadata(task_result).unwrap_or_else(Vec::new),
-    }
-  }
+  pub(crate) fn update_workunit(workunit: &mut RunningWorkunit, task_result: &PyAny) {
+    workunit.update_metadata(|old_metadata| {
+      let new_level = Self::level(task_result);
+      // If the metadata already existed, or if its level changed, we need to update it.
+      let mut metadata = if new_level.is_some() || old_metadata.is_some() {
+        old_metadata.unwrap_or_default()
+      } else {
+        return None;
+      };
 
-  pub(crate) fn update_workunit(self, workunit: &mut RunningWorkunit) {
-    workunit.update_metadata(|mut metadata| {
-      if let Some(new_level) = self.level {
+      if let Some(new_level) = new_level {
         metadata.level = new_level;
       }
-      metadata.message = self.message;
-      metadata.artifacts.extend(self.artifacts);
-      metadata.user_metadata.extend(self.metadata);
+      metadata.message = Self::message(task_result);
       metadata
+        .artifacts
+        .extend(Self::artifacts(task_result).unwrap_or_else(Vec::new));
+      metadata
+        .user_metadata
+        .extend(metadata_for(task_result).unwrap_or_else(Vec::new));
+      Some(metadata)
     });
   }
 
@@ -103,11 +101,11 @@ impl EngineAwareParameter {
   }
 
   pub fn metadata(obj: &PyAny) -> Vec<(String, UserMetadataItem)> {
-    metadata(obj).unwrap_or_else(Vec::new)
+    metadata_for(obj).unwrap_or_else(Vec::new)
   }
 }
 
-fn metadata(obj: &PyAny) -> Option<Vec<(String, UserMetadataItem)>> {
+fn metadata_for(obj: &PyAny) -> Option<Vec<(String, UserMetadataItem)>> {
   let metadata_val = obj.call_method0("metadata").ok()?;
   if metadata_val.is_none() {
     return None;

--- a/src/rust/engine/src/externs/interface.rs
+++ b/src/rust/engine/src/externs/interface.rs
@@ -667,6 +667,15 @@ async fn workunit_to_py_value(
   workunit: Workunit,
   core: &Arc<Core>,
 ) -> PyO3Result<Value> {
+  let metadata = workunit.metadata.ok_or_else(|| {
+    PyException::new_err(format!(
+      // TODO: It would be better for this to be typesafe, but it isn't currently worth it to
+      // split the Workunit struct.
+      "Workunit for {} was disabled. Please file an issue at \
+            [https://github.com/pantsbuild/pants/issues/new].",
+      workunit.span_id
+    ))
+  })?;
   let has_parent_ids = !workunit.parent_ids.is_empty();
   let mut dict_entries = {
     let gil = Python::acquire_gil();
@@ -682,7 +691,7 @@ async fn workunit_to_py_value(
       ),
       (
         externs::store_utf8(py, "level"),
-        externs::store_utf8(py, &workunit.metadata.level.to_string()),
+        externs::store_utf8(py, &metadata.level.to_string()),
       ),
     ];
 
@@ -739,7 +748,7 @@ async fn workunit_to_py_value(
       }
     };
 
-    if let Some(desc) = &workunit.metadata.desc.as_ref() {
+    if let Some(desc) = &metadata.desc.as_ref() {
       dict_entries.push((
         externs::store_utf8(py, "description"),
         externs::store_utf8(py, desc),
@@ -750,7 +759,7 @@ async fn workunit_to_py_value(
 
   let mut artifact_entries = Vec::new();
 
-  for (artifact_name, digest) in workunit.metadata.artifacts.iter() {
+  for (artifact_name, digest) in metadata.artifacts.iter() {
     let store = core.store();
     let py_val = match digest {
       ArtifactOutput::FileDigest(digest) => {
@@ -786,8 +795,8 @@ async fn workunit_to_py_value(
   let gil = Python::acquire_gil();
   let py = gil.python();
 
-  let mut user_metadata_entries = Vec::with_capacity(workunit.metadata.user_metadata.len());
-  for (user_metadata_key, user_metadata_item) in workunit.metadata.user_metadata.iter() {
+  let mut user_metadata_entries = Vec::with_capacity(metadata.user_metadata.len());
+  for (user_metadata_key, user_metadata_item) in metadata.user_metadata.iter() {
     let value = match user_metadata_item {
       UserMetadataItem::ImmediateString(v) => v.into_py(py),
       UserMetadataItem::ImmediateInt(n) => n.into_py(py),
@@ -810,14 +819,14 @@ async fn workunit_to_py_value(
     externs::store_dict(py, user_metadata_entries)?,
   ));
 
-  if let Some(stdout_digest) = workunit.metadata.stdout {
+  if let Some(stdout_digest) = metadata.stdout {
     artifact_entries.push((
       externs::store_utf8(py, "stdout_digest"),
       crate::nodes::Snapshot::store_file_digest(py, stdout_digest).map_err(PyException::new_err)?,
     ));
   }
 
-  if let Some(stderr_digest) = workunit.metadata.stderr {
+  if let Some(stderr_digest) = metadata.stderr {
     artifact_entries.push((
       externs::store_utf8(py, "stderr_digest"),
       crate::nodes::Snapshot::store_file_digest(py, stderr_digest).map_err(PyException::new_err)?,

--- a/src/rust/engine/src/nodes.rs
+++ b/src/rust/engine/src/nodes.rs
@@ -433,24 +433,26 @@ impl WrappedNode for ExecuteProcess {
 
     let definition = serde_json::to_string(&request)
       .map_err(|e| throw(format!("Failed to serialize process: {}", e)))?;
-    workunit.update_metadata(|initial| WorkunitMetadata {
-      stdout: Some(res.stdout_digest),
-      stderr: Some(res.stderr_digest),
-      user_metadata: vec![
-        (
-          "definition".to_string(),
-          UserMetadataItem::ImmediateString(definition),
-        ),
-        (
-          "source".to_string(),
-          UserMetadataItem::ImmediateString(format!("{:?}", res.metadata.source)),
-        ),
-        (
-          "exit_code".to_string(),
-          UserMetadataItem::ImmediateInt(res.exit_code as i64),
-        ),
-      ],
-      ..initial
+    workunit.update_metadata(|initial| {
+      initial.map(|initial| WorkunitMetadata {
+        stdout: Some(res.stdout_digest),
+        stderr: Some(res.stderr_digest),
+        user_metadata: vec![
+          (
+            "definition".to_string(),
+            UserMetadataItem::ImmediateString(definition),
+          ),
+          (
+            "source".to_string(),
+            UserMetadataItem::ImmediateString(format!("{:?}", res.metadata.source)),
+          ),
+          (
+            "exit_code".to_string(),
+            UserMetadataItem::ImmediateInt(res.exit_code as i64),
+          ),
+        ],
+        ..initial
+      })
     });
     if let Some(total_elapsed) = res.metadata.total_elapsed {
       let total_elapsed = Duration::from(total_elapsed).as_millis() as u64;
@@ -1183,14 +1185,11 @@ impl WrappedNode for Task {
       )));
     }
 
-    let engine_aware_return_type = if self.task.engine_aware_return_type {
+    if self.task.engine_aware_return_type {
       let gil = Python::acquire_gil();
       let py = gil.python();
-      EngineAwareReturnType::from_task_result((*result_val).as_ref(py))
-    } else {
-      EngineAwareReturnType::default()
+      EngineAwareReturnType::update_workunit(workunit, (*result_val).as_ref(py))
     };
-    engine_aware_return_type.update_workunit(workunit);
 
     Ok(result_val)
   }

--- a/src/rust/engine/workunit_store/Cargo.toml
+++ b/src/rust/engine/workunit_store/Cargo.toml
@@ -22,4 +22,6 @@ strum_macros = "0.23"
 tokio = { version = "1.16", features = ["rt", "sync"] }
 
 [dev-dependencies]
+futures = "0.3"
 internment = "0.6"
+tokio = { version = "1.16", features = ["macros"] }

--- a/src/rust/engine/workunit_store/src/lib.rs
+++ b/src/rust/engine/workunit_store/src/lib.rs
@@ -48,8 +48,7 @@ use petgraph::visit::{VisitMap, Visitable};
 use rand::thread_rng;
 use rand::Rng;
 use smallvec::SmallVec;
-use tokio::sync::broadcast::error::TryRecvError;
-use tokio::sync::broadcast::{self, Receiver, Sender};
+use tokio::sync::mpsc::{self, UnboundedReceiver, UnboundedSender};
 use tokio::task_local;
 
 mod metrics;
@@ -380,19 +379,19 @@ enum StoreMsg {
 pub struct WorkunitStore {
   log_starting_workunits: bool,
   max_level: Level,
-  sender: Sender<StoreMsg>,
+  senders: [UnboundedSender<StoreMsg>; 2],
   streaming_workunit_data: Arc<Mutex<StreamingWorkunitData>>,
   heavy_hitters_data: Arc<Mutex<HeavyHittersData>>,
   metrics_data: Arc<MetricsData>,
 }
 
 struct StreamingWorkunitData {
-  receiver: Receiver<StoreMsg>,
+  receiver: UnboundedReceiver<StoreMsg>,
   running_graph: RunningWorkunitGraph,
 }
 
 impl StreamingWorkunitData {
-  fn new(receiver: Receiver<StoreMsg>) -> StreamingWorkunitData {
+  fn new(receiver: UnboundedReceiver<StoreMsg>) -> StreamingWorkunitData {
     StreamingWorkunitData {
       receiver,
       running_graph: RunningWorkunitGraph::default(),
@@ -404,20 +403,7 @@ impl StreamingWorkunitData {
 
     let mut started_workunits = Vec::new();
     let mut completed_workunits = Vec::new();
-    loop {
-      let msg = match self.receiver.try_recv() {
-        Ok(msg) => msg,
-        Err(TryRecvError::Closed | TryRecvError::Empty) => break,
-        Err(TryRecvError::Lagged(skipped)) => {
-          log::warn!(
-            "The StreamingWorkunitHandler fell behind on workunit processing: \
-            {skipped} messages were dropped.\n\
-            Try lowering the `--streaming-workunits-report-interval` to ensure that workunits \
-            are processed in a timely manner."
-          );
-          continue;
-        }
-      };
+    while let Ok(msg) = self.receiver.try_recv() {
       match msg {
         StoreMsg::Started(mut started) => {
           self.running_graph.add(started.clone());
@@ -452,12 +438,12 @@ impl StreamingWorkunitData {
 }
 
 struct HeavyHittersData {
-  receiver: Receiver<StoreMsg>,
+  receiver: UnboundedReceiver<StoreMsg>,
   running_graph: RunningWorkunitGraph,
 }
 
 impl HeavyHittersData {
-  fn new(receiver: Receiver<StoreMsg>) -> HeavyHittersData {
+  fn new(receiver: UnboundedReceiver<StoreMsg>) -> HeavyHittersData {
     HeavyHittersData {
       receiver,
       running_graph: RunningWorkunitGraph::default(),
@@ -465,20 +451,7 @@ impl HeavyHittersData {
   }
 
   fn refresh_store(&mut self) {
-    loop {
-      let msg = match self.receiver.try_recv() {
-        Ok(msg) => msg,
-        Err(TryRecvError::Closed | TryRecvError::Empty) => break,
-        Err(TryRecvError::Lagged(skipped)) => {
-          log::warn!(
-            "The `--dynamic-ui` fell behind on workunit processing: \
-            {skipped} messages were dropped.\n\
-            This is unexpected: please file an issue at \
-            `https://github.com/pantsbuild/pants/issues/new/choose`."
-          );
-          continue;
-        }
-      };
+    while let Ok(msg) = self.receiver.try_recv() {
       match msg {
         StoreMsg::Started(started) => self.running_graph.add(started),
         StoreMsg::Completed(span_id, new_metadata, time) => {
@@ -594,17 +567,18 @@ impl HeavyHittersData {
 
 impl WorkunitStore {
   pub fn new(log_starting_workunits: bool, max_level: Level) -> WorkunitStore {
-    // NB: This is a relatively large allocation. The UI will poll multiple times a second and
-    // shouldn't fall behind. But the streaming workunit subscriber has a configurable poll
-    // frequency, and the error message below suggests adjusting that if messages are dropped.
-    let (sender, receiver1) = broadcast::channel(16384);
-    let receiver2 = sender.subscribe();
+    // NB: Although it would be nice not to have seperate allocations per consumer, it is
+    // difficult to use a channel like `tokio::sync::broadcast` due to that channel being bounded.
+    // Subscribers receive messages at very different rates, and adjusting the workunit level
+    // affects the total number of messages that might be queued at any given time.
+    let (sender1, receiver1) = mpsc::unbounded_channel();
+    let (sender2, receiver2) = mpsc::unbounded_channel();
     WorkunitStore {
       log_starting_workunits,
       max_level,
       // TODO: Create one `StreamingWorkunitData` per subscriber, and zero if no subscribers are
       // installed.
-      sender,
+      senders: [sender1, sender2],
       streaming_workunit_data: Arc::new(Mutex::new(StreamingWorkunitData::new(receiver1))),
       heavy_hitters_data: Arc::new(Mutex::new(HeavyHittersData::new(receiver2))),
       metrics_data: Arc::default(),
@@ -641,6 +615,19 @@ impl WorkunitStore {
     self.heavy_hitters_data.lock().heavy_hitters(k)
   }
 
+  fn send(&self, msg: StoreMsg) {
+    let send_inner = |sender: &UnboundedSender<StoreMsg>, msg: StoreMsg| {
+      sender
+        .send(msg)
+        .unwrap_or_else(|_| panic!("Receivers are static, and should always be present."));
+    };
+    // Send clones to the first N-1 senders, and the owned value to the final sender.
+    for sender in &self.senders[0..self.senders.len() - 1] {
+      send_inner(sender, msg.clone());
+    }
+    send_inner(&self.senders[self.senders.len() - 1], msg);
+  }
+
   ///
   /// NB: Public for macro use. Use `in_workunit!` instead.
   ///
@@ -662,10 +649,7 @@ impl WorkunitStore {
       metadata,
     };
 
-    self
-      .sender
-      .send(StoreMsg::Started(started.clone()))
-      .unwrap_or_else(|_| panic!("Receivers are static, and should always be present."));
+    self.send(StoreMsg::Started(started.clone()));
 
     if self.log_starting_workunits {
       started.log_workunit_state(false)
@@ -679,23 +663,17 @@ impl WorkunitStore {
 
   fn cancel_workunit(&self, workunit: Workunit) {
     workunit.log_workunit_state(true);
-    self
-      .sender
-      .send(StoreMsg::Canceled(
-        workunit.span_id,
-        std::time::SystemTime::now(),
-      ))
-      .unwrap_or_else(|_| panic!("Receivers are static, and should always be present."));
+    self.send(StoreMsg::Canceled(
+      workunit.span_id,
+      std::time::SystemTime::now(),
+    ));
   }
 
   fn complete_workunit_impl(&self, mut workunit: Workunit, end_time: SystemTime) {
     let span_id = workunit.span_id;
     let new_metadata = workunit.metadata.clone();
 
-    self
-      .sender
-      .send(StoreMsg::Completed(span_id, new_metadata, end_time))
-      .unwrap_or_else(|_| panic!("Receivers are static, and should always be present."));
+    self.send(StoreMsg::Completed(span_id, new_metadata, end_time));
 
     let start_time = match workunit.state {
       WorkunitState::Started { start_time, .. } => start_time,
@@ -731,11 +709,7 @@ impl WorkunitStore {
       metadata: Some(metadata),
     };
 
-    self
-      .sender
-      .send(StoreMsg::Started(workunit.clone()))
-      .unwrap_or_else(|_| panic!("Receivers are static, and should always be present."));
-
+    self.send(StoreMsg::Started(workunit.clone()));
     self.complete_workunit_impl(workunit, end_time);
   }
 

--- a/src/rust/engine/workunit_store/src/lib.rs
+++ b/src/rust/engine/workunit_store/src/lib.rs
@@ -131,9 +131,7 @@ impl RunningWorkunitGraph {
             workunit.state = WorkunitState::Completed { time_span };
           }
         };
-        if let Some(metadata) = new_metadata {
-          workunit.metadata = metadata;
-        }
+        workunit.metadata = new_metadata;
         Some(workunit)
       }
     }
@@ -213,6 +211,17 @@ impl RunningWorkunitGraph {
 /// When the `in_workunit!` macro exits, the Workunit on the stack is completed by storing any
 /// local mutated values as the final value of the Workunit.
 ///
+/// NB: A Workunit which has too low a level for the WorkunitStore is called "disabled", but is
+/// still recorded (without any metadata). This is for a few reasons:
+///   1. It can be important to have accurate transitive parents when workunits are treated as a
+///      DAG (see #14680): otherwise, adding an additional parent to an existing workunit is
+///      tricky, because granularity is lost. For example: if `A -> B -> C` are workunits, and `B`
+///      is disabled, an attempt to add a new parent to `B` would not have a record of the
+///      relationship to `C`.
+///   2. When a workunit completes, it is able to adjust its level (if it errored, for example),
+///      and it is cleaner semantically to re-enable a disabled workunit when it completes than
+///      to special case the creation of a the workunit at that point.
+///
 #[derive(Clone, Debug)]
 pub struct Workunit {
   pub name: &'static str,
@@ -221,22 +230,32 @@ pub struct Workunit {
   // it may gain additional parents due to memoization.
   pub parent_ids: ParentIds,
   pub state: WorkunitState,
-  pub metadata: WorkunitMetadata,
+  pub metadata: Option<WorkunitMetadata>,
 }
 
 impl Workunit {
+  fn enabled_at(&self, level: Level) -> bool {
+    self
+      .metadata
+      .as_ref()
+      .map(|m| m.level <= level)
+      .unwrap_or(false)
+  }
+
   fn log_workunit_state(&self, canceled: bool) {
-    if !log::log_enabled!(self.metadata.level) {
-      return;
-    }
+    let metadata = match self.metadata.as_ref() {
+      Some(metadata) if log::log_enabled!(metadata.level) => metadata,
+      _ => return,
+    };
+
     let state = match (&self.state, canceled) {
       (_, true) => "Canceled:",
       (WorkunitState::Started { .. }, _) => "Starting:",
       (WorkunitState::Completed { .. }, _) => "Completed:",
     };
 
-    let level = self.metadata.level;
-    let identifier = if let Some(ref s) = self.metadata.desc {
+    let level = metadata.level;
+    let identifier = if let Some(ref s) = metadata.desc {
       s.as_str()
     } else {
       self.name
@@ -260,7 +279,7 @@ impl Workunit {
       identifier.to_string()
     };
 
-    let message = if let Some(ref s) = self.metadata.message {
+    let message = if let Some(ref s) = metadata.message {
       format!(" - {}", s)
     } else {
       "".to_string()
@@ -381,7 +400,7 @@ impl StreamingWorkunitData {
   }
 
   pub fn latest_workunits(&mut self, max_verbosity: log::Level) -> (Vec<Workunit>, Vec<Workunit>) {
-    let should_emit = |workunit: &Workunit| -> bool { workunit.metadata.level <= max_verbosity };
+    let should_emit = |workunit: &Workunit| -> bool { workunit.enabled_at(max_verbosity) };
 
     let mut started_workunits = Vec::new();
     let mut completed_workunits = Vec::new();
@@ -496,7 +515,7 @@ impl HeavyHittersData {
     let mut res = HashMap::new();
     while let Some((Reverse(start_time), span_id)) = queue.pop() {
       let (_, workunit) = self.running_graph.entries.get(&span_id).unwrap();
-      if let Some(effective_name) = workunit.metadata.desc.as_ref() {
+      if let Some(effective_name) = workunit.metadata.as_ref().and_then(|m| m.desc.as_ref()) {
         res.insert(workunit.span_id, (effective_name.to_string(), start_time));
         if res.len() >= k {
           break;
@@ -524,7 +543,14 @@ impl HeavyHittersData {
       .filter_map(|(_, workunit)| {
         let duration = Self::duration_for(now, workunit)?;
         if duration >= duration_threshold {
-          Some((workunit.metadata.desc.as_ref()?.clone(), duration))
+          Some((
+            workunit
+              .metadata
+              .as_ref()
+              .and_then(|m| m.desc.as_ref())?
+              .clone(),
+            duration,
+          ))
         } else {
           None
         }
@@ -545,8 +571,12 @@ impl HeavyHittersData {
   }
 
   fn is_visible(workunit: &Workunit) -> bool {
-    workunit.metadata.level <= Level::Debug
-      && workunit.metadata.desc.is_some()
+    workunit.enabled_at(Level::Debug)
+      && workunit
+        .metadata
+        .as_ref()
+        .and_then(|m| m.desc.as_ref())
+        .is_some()
       && matches!(workunit.state, WorkunitState::Started { .. })
   }
 
@@ -619,7 +649,7 @@ impl WorkunitStore {
     span_id: SpanId,
     name: &'static str,
     parent_id: Option<SpanId>,
-    metadata: WorkunitMetadata,
+    metadata: Option<WorkunitMetadata>,
   ) -> Workunit {
     let started = Workunit {
       name,
@@ -660,7 +690,7 @@ impl WorkunitStore {
 
   fn complete_workunit_impl(&self, mut workunit: Workunit, end_time: SystemTime) {
     let span_id = workunit.span_id;
-    let new_metadata = Some(workunit.metadata.clone());
+    let new_metadata = workunit.metadata.clone();
 
     self
       .sender
@@ -698,7 +728,7 @@ impl WorkunitStore {
         start_time,
         blocked: Arc::new(AtomicBool::new(false)),
       },
-      metadata,
+      metadata: Some(metadata),
     };
 
     self
@@ -778,8 +808,8 @@ impl WorkunitStore {
   pub fn setup_for_tests() -> (WorkunitStore, RunningWorkunit) {
     let store = WorkunitStore::new(false, Level::Trace);
     store.init_thread_state(None);
-    let workunit = store._start_workunit(SpanId(0), "testing", None, WorkunitMetadata::default());
-    (store.clone(), RunningWorkunit::new(store, Some(workunit)))
+    let workunit = store._start_workunit(SpanId(0), "testing", None, Option::default());
+    (store.clone(), RunningWorkunit::new(store, workunit))
   }
 }
 
@@ -845,44 +875,37 @@ macro_rules! in_workunit {
     use futures::future::FutureExt;
     let mut store_handle = $crate::expect_workunit_store_handle();
     let level: log::Level  = $workunit_level;
-    if store_handle.store.max_level() >= level {
-      let mut $workunit = {
-        let workunit_metadata = $crate::WorkunitMetadata {
-          level,
-          $(
-                $workunit_field_name: $workunit_field_value,
-          )*
-          ..Default::default()
+    let mut $workunit = {
+      let workunit_metadata =
+        if store_handle.store.max_level() >= level {
+          Some($crate::WorkunitMetadata {
+            level,
+            $(
+                  $workunit_field_name: $workunit_field_value,
+            )*
+            ..Default::default()
+          })
+        } else {
+          None
         };
-        let span_id = $crate::SpanId::new();
-        let parent_id = std::mem::replace(&mut store_handle.parent_id, Some(span_id));
-        let workunit =
-          store_handle
-            .store
-            ._start_workunit(span_id, $workunit_name, parent_id, workunit_metadata);
-        $crate::RunningWorkunit::new(store_handle.store.clone(), Some(workunit))
-      };
-      $crate::scope_task_workunit_store_handle(Some(store_handle), async move {
-        let result = {
-          let $workunit = &mut $workunit;
-          $f
-        }
-        .await;
-        $workunit.complete();
-        result
-      })
-      .boxed()
-    } else {
-      async move {
-        let mut $workunit = $crate::RunningWorkunit::new(store_handle.store, None);
-        {
-          let $workunit = &mut $workunit;
-          $f
-        }
-        .await
+      let span_id = $crate::SpanId::new();
+      let parent_id = std::mem::replace(&mut store_handle.parent_id, Some(span_id));
+      let workunit =
+        store_handle
+          .store
+          ._start_workunit(span_id, $workunit_name, parent_id, workunit_metadata);
+      $crate::RunningWorkunit::new(store_handle.store.clone(), workunit)
+    };
+    $crate::scope_task_workunit_store_handle(Some(store_handle), async move {
+      let result = {
+        let $workunit = &mut $workunit;
+        $f
       }
-      .boxed()
-    }
+      .await;
+      $workunit.complete();
+      result
+    })
+    .boxed()
   }};
 }
 
@@ -892,8 +915,11 @@ pub struct RunningWorkunit {
 }
 
 impl RunningWorkunit {
-  pub fn new(store: WorkunitStore, workunit: Option<Workunit>) -> RunningWorkunit {
-    RunningWorkunit { store, workunit }
+  pub fn new(store: WorkunitStore, workunit: Workunit) -> RunningWorkunit {
+    RunningWorkunit {
+      store,
+      workunit: Some(workunit),
+    }
   }
 
   pub fn record_observation(&mut self, metric: ObservationMetric, value: u64) {
@@ -904,9 +930,14 @@ impl RunningWorkunit {
     self.store.increment_counter(counter_name, change);
   }
 
+  ///
+  /// If the workunit is enabled, receives its current metadata. If Some(metadata) is returned by
+  /// this method for a disabled workunit, the workunit will complete as enabled if the Level set
+  /// by the new metadata is high enough to re-enable it.
+  ///
   pub fn update_metadata<F>(&mut self, f: F)
   where
-    F: FnOnce(WorkunitMetadata) -> WorkunitMetadata,
+    F: FnOnce(Option<WorkunitMetadata>) -> Option<WorkunitMetadata>,
   {
     if let Some(ref mut workunit) = self.workunit {
       workunit.metadata = f(workunit.metadata.clone())

--- a/src/rust/engine/workunit_store/src/tests.rs
+++ b/src/rust/engine/workunit_store/src/tests.rs
@@ -4,7 +4,7 @@ use std::time::Duration;
 
 use internment::Intern;
 
-use crate::{SpanId, WorkunitMetadata, WorkunitState, WorkunitStore};
+use crate::{Level, SpanId, WorkunitMetadata, WorkunitState, WorkunitStore};
 
 #[test]
 fn heavy_hitters_basic() {
@@ -56,6 +56,46 @@ fn straggling_workunits_blocked_path() {
   // Test that a chain of blocked workunits do not cause their parents to be rendered.
   let ws = create_store(vec![wu_root(0)], vec![wu(1, 0), wu(2, 1)], vec![]);
   assert!(ws.straggling_workunits(Duration::from_secs(0)).is_empty());
+}
+
+#[tokio::test]
+async fn workunit_escalation_is_recorded() {
+  // Create a store which will disable Debug level workunits.
+  let ws = WorkunitStore::new(true, Level::Info);
+  ws.init_thread_state(None);
+
+  // Start a workunit at Debug (below the level of the store).
+  let new_desc = "One more thing!";
+  in_workunit!(
+    "super_fine",
+    Level::Debug,
+    desc = Some("Should be ignored".to_owned()),
+    |workunit| async move {
+      workunit.update_metadata(|metadata| {
+        // Ensure that it has no metadata (i.e.: is disabled).
+        assert!(metadata.is_none());
+
+        // Then return new metadata to raise the workunit's level.
+        Some(WorkunitMetadata {
+          level: Level::Info,
+          desc: Some(new_desc.to_owned()),
+          ..WorkunitMetadata::default()
+        })
+      });
+    }
+  )
+  .await;
+
+  // Finally, confirm that the workunit did end up recorded using the new level.
+  let (started, completed) = ws.latest_workunits(Level::Info);
+  assert!(started.is_empty());
+  assert_eq!(
+    completed
+      .into_iter()
+      .map(|wu| wu.metadata.and_then(|m| m.desc))
+      .collect::<Vec<_>>(),
+    vec![Some(new_desc.to_owned())]
+  );
 }
 
 #[test]
@@ -115,7 +155,7 @@ fn create_store(
         span_id,
         Intern::new(format!("{}", span_id.0)).as_ref(),
         parent_id,
-        metadata,
+        Some(metadata),
       )
     })
     .collect::<Vec<_>>();

